### ISSUE TITLE
Strange behavior for tags after \clip inside \t

### ISSUE THREAD

**Issue by SSgumS** [libass/libass#472]
When using \clip in \t, tags after it will apply from the beginning and doesn't respect the timing specified for \t. To get a workaround I must keep \clip the last tag.
Example:
![image](https://user-images.githubusercontent.com/25867589/105753991-c153ef00-5f5e-11eb-8e7e-a73e2bdff3a0.png)
This line must be yellow at first and gradually turn to purple. But as you can see, it is purple from the beginning. It gets fixed when I switch occurrence order of \clip and \1c.

-----
libass version: [This Commit](https://github.com/libass/libass/commit/7a853255bb5083c9c6ab31a29f215875dc8d8251)

**Comment by astiob** [comment]
Yes, this is how ASS works in every renderer except MPC-HC. If you rely on the opposite behaviour, your file will not play correctly in more than just libass. And if libass were to implement the opposite behaviour, other files would start to render incorrectly in libass.

**Comment by astiob** [comment]
I was reminded that the last releases of XySubFilter also support nested parentheses (by way of a parsing code merge from MPC-HC). On the other hand, xy-VSFilter and older VSFilters don’t (and handle this like current libass does). If there is evidence that nested parentheses are used in the wild more than the nesting-unsupported syntax is, I’m ready to revisit this. Personally, I _hope_ that very few files rely on _either_ parsing, as they are inevitably misparsed in some of the renderers. But when in doubt, I prefer to stick to older behaviour models in hopes of correctly rendering more files.

**Comment by SSgumS** [comment]
Thank you for your explanation.
I think it is so simple for people to pair each starting parenthesis with a balanced ending one. We always use parentheses in a pair. So I don't think there would be much people that use this irrational behavior on purpose. Usually human beings avoid things that are not logical.
Maybe at the beginning of ASS, it was hard to implement a balanced parentheses parser, so they decided to stick this behavior.
It is never too late to mend.

---
Aegisub's default subtitle render is VSFilter and recent builds of it has a version of VSFilter that parses parentheses in pairs. If we stick to the old non-balanced behavior, we may ruin new subs. [As Aegisub is subbers' standard, most of them may think that their sub is OK (and they don't doubt it because the behavior is rational and Aegisub is showing their sub fine). But they don't have a clue that libass doesn't support their subs]
In addition, some popular players in Windows like PotPlayer and MPC-HC are supporting this too.

**Comment by astiob** [comment]
Update that I keep forgetting to post:

> Aegisub's default subtitle render is VSFilter and recent builds of it has a version of VSFilter that parses parentheses in pairs.

If you mean https://github.com/Ristellise/AegisubDC/releases, this was an unintended mistake and has now been fixed (and the affected releases haven’t been out for long). Keep in mind these Aegisub builds also include VSFilterMod, which does parse parentheses in pairs.

> Maybe at the beginning of ASS, it was hard to implement a balanced parentheses parser, so they decided to stick this behavior.

(it isn’t hard, but nobody thought about it; or possibly, though less likely, SubStation Alpha didn’t do it and VSFilter followed that)

> As Aegisub is subbers' standard, most of them may think that their sub is OK (and they don't doubt it because the behavior is rational and Aegisub is showing their sub fine).

This is precisely the kind of reason why we want to preserve traditional VSFilter compatibility, including parenthesis matching. And why people do use a _whole lot_ of irrational things in ASS, and why it _is_ too late to mend them without breaking existing files: they don’t care what’s rational, but rather what works.

I’ve asked a person with a huge database of ASS scripts (although I’m not sure how much it’s biased towards English subtitles) to run a search for files that would be affected by parentheses matching changes. We haven’t (yet) been able to formulate the search precisely enough to get reliable statistics out of it, but based on a sample of 1000 files that I was shown, we found _some_ files that rely on unbalanced parentheses. It wasn’t terribly common, and the visual differences weren’t huge, and some of those files seem to have gotten their parentheses balanced out in later rereleases. But at the same time, we found _zero_ files that rely on balanced nested parentheses.

For what it’s worth, note also that VSFilterMod’s/MPC-HC’s parenthesis parsing is not fully logical, either. For example, nested parentheses are parsed in any arguments (not just in the last argument of \t), and they can span _across_ several arguments. But in its defence, it has one indisputable benefit that perhaps nobody thought of: it allows using `\fn` and `\r` with fonts and styles whose names contain parentheses, which is simply impossible in guliverkli/xy-VSFilter/libass.
